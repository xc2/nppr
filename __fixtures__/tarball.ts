export const BasicTarballPath = new URL("./barhop-0.0.0-PLACEHOLDER.tgz", import.meta.url).pathname;
