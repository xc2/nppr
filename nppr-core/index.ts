export { tryToNumber } from "./internals/lang";
export { duplicate } from "./internals/stream";
export {
  inlineTemplate,
  extractName,
  extractVersion,
  getPackageTokens,
  template,
  inputSource,
  toPurl,
  type InputSource,
  NPPR_USER_AGENT,
} from "./utils";
export { type Manifest, getManifest } from "./internals/package";
