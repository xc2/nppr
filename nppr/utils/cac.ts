import { EOL } from "node:os";
import * as process from "node:process";
import type { CAC, Command } from "cac";
import stringWidth from "string-width";

export interface CliCommand {
  (get: CAC["command"]): Command;
}

export const registerCommand = (cli: CAC, name: string, cmd: CliCommand) => {
  const register: CAC["command"] = (rawName, ...args) => {
    return cli.command([name, rawName].filter(Boolean).join(" "), ...args);
  };
  return cmd(register);
};

export class ArgumentsError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ArgumentsError";
  }
}

export function getOptionsForUsage(cli: CAC) {
  const cmd = cli.matchedCommand || cli.globalCommand;
  const { options: globalOptions } = cli.globalCommand;
  let options = cmd.isGlobalCommand ? globalOptions : [...cmd.options, ...(globalOptions || [])];

  if (!cmd.isGlobalCommand && !cmd.isDefaultCommand) {
    options = options.filter((option) => option.name !== "version");
  }

  const maxColumn = getMaxLength(options.map((option) => option.rawName));
  return {
    maxColumn,
    options,
  };
}

function getMaxLength(arr: string[]) {
  return arr.reduce((a, b) => Math.max(a, b.length), 0);
}

export function indent(str: string, count = 2) {
  const spaces = " ".repeat(count);
  return str.replace(/^(?=.+$)/gm, spaces);
}

export function trimLeft(str: string) {
  return str.replace(/^\s+/g, "");
}

export function simplyFold(str: string, ind = 0, columns?: number) {
  const renderColumns = (columns || process.stdout.columns || 80) - ind;
  const lines = str.split(/(\r\n|\r|\n)/g);
  const spaces = " ".repeat(ind);
  const fin: string[] = [];
  for (let i = 0; i < lines.length; i += 2) {
    const line = lines[i];
    const br = lines[i + 1] ?? "";
    const words = line.split(/(\s+)/g);
    let currentLine = "";
    let currentLength = -1;
    for (let j = 0; j < words.length; j += 2) {
      const word = words[j] + (words[j + 1] ?? "");
      const len = stringWidth(word);
      const nextLen = currentLength + len;
      if (nextLen > renderColumns) {
        fin.push(spaces + currentLine + EOL);
        currentLine = word;
        currentLength = len;
      } else {
        currentLine += word;
        currentLength = nextLen;
      }
    }
    if (currentLine) {
      fin.push(spaces + currentLine);
    }
    fin.push(br);
  }

  return fin.join("");
}
