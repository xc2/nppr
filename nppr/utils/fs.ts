import { createReadStream, createWriteStream } from "node:fs";
import { tryToNumber } from "./lang";

export function outputStream(filepath: string | number) {
  filepath = tryToNumber(filepath);

  if (typeof filepath === "string") {
    // file path
    return createWriteStream(filepath);
  }
  if (typeof filepath === "number") {
    const builtins: Record<string, NodeJS.WritableStream> = {
      [process.stdout.fd]: process.stdout,
      [process.stderr.fd]: process.stderr,
    };
    // file descriptor
    return builtins[filepath] ?? createReadStream("", { fd: filepath });
  }
  throw new Error(`Invalid output filepath: ${filepath}`);
}
