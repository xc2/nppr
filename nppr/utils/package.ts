import { createWriteStream } from "node:fs";
import * as NodePath from "node:path";
import { Writable } from "node:stream";
import { type Manifest, getManifest, inlineTemplate, inputSource, tryToNumber } from "nppr-core";

export function pathInfo(path: string, cwd: string) {
  const abs = NodePath.resolve(cwd, path);
  const rel = NodePath.relative(cwd, abs);
  let ext = NodePath.extname(abs);
  let base = NodePath.basename(abs, ext);
  if (NodePath.extname(base) === ".tar") {
    ext = ".tar" + ext;
    base = NodePath.basename(base, ext);
  }
  return { fullpath: abs, path: NodePath.dirname(rel), extname: ext, basename: base };
}

export function packageName(name: string) {
  if (name.startsWith("@")) {
    const [scope, unscoped] = name.slice(1).split("/");
    return { scope, unscoped, pathPart: `${scope}-${unscoped}` };
  }
  return { scope: "", unscoped: name, pathPart: name };
}

export class Package {
  public readonly input: string | number;
  public readonly pathInfo?: ReturnType<typeof pathInfo>;
  protected _source?: ReadableStream;
  protected _manifest?: Promise<Manifest>;

  constructor(
    input: string | number,
    public readonly cwd = process.cwd()
  ) {
    const p = tryToNumber(input);
    if (typeof p === "string") {
      this.pathInfo = pathInfo(p, this.cwd);
      this.input = this.pathInfo.fullpath;
    } else {
      this.input = p;
    }
  }

  tee() {
    const [a, b] = this.source.tee();
    this._source = a;
    return b;
  }

  get source() {
    this._source = this._source ?? inputSource(this.input);
    return this._source;
  }

  set source(source: ReadableStream) {
    if (this._source !== source) {
      this._source = source;
      this._manifest = undefined;
    }
  }

  get manifest() {
    try {
      this._manifest = this._manifest || getManifest(this.tee());
      return this._manifest;
    } catch (e) {
      return Promise.reject(e);
    }
  }

  async output(p?: string) {
    const outputPath = await this.getOutputPath(p);
    return this.tee().pipeTo(Writable.toWeb(createWriteStream(outputPath)));
  }

  // TODO: move along and add tests
  async getOutputPath(p: string = "[path]/[name]-[version][extname]") {
    const manifest = await this.manifest;
    const name = packageName(manifest.name);
    const variables = {
      ...manifest,
      scope: name.scope,
      unscoped: name.unscoped,
      fullpath: this.pathInfo?.fullpath ?? "",
      path: this.pathInfo?.path ?? "",
      extname: this.pathInfo?.extname ?? ".tgz",
      basename: this.pathInfo?.basename ?? `${name.pathPart}-${manifest.version}`,
    };
    const _escape = (value: any, key: string) => {
      if (key === "name") {
        return packageName(value).pathPart;
      }
      return `${value}`.replace(/[^0-9a-zA-Z-._]/g, "_");
    };
    return NodePath.resolve(this.cwd, inlineTemplate(p, { escape: _escape })(variables));
  }
}
