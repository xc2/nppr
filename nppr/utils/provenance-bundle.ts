import { readFile, writeFile } from "node:fs/promises";
import { type Manifest, toPurl } from "nppr-core";

export class ProvenanceBundle {
  protected readonly map: Map<string, any> = new Map();

  add(provenance: any) {
    const { subject } = JSON.parse(
      Buffer.from(provenance.dsseEnvelope.payload, "base64").toString("utf-8")
    );
    // TODO validate subject for NPM
    this.map.set(subject[0].name, provenance);
  }

  get(name: string | Pick<Manifest, "name" | "version">) {
    return this.map.get(typeof name === "string" ? name : toPurl(name));
  }

  async outputBundle(path: string) {
    const content = Array.from(this.map.values(), (v) => JSON.stringify(v)).join("\n");
    await writeFile(path, content);
  }
  async outputSingle(name: string | Pick<Manifest, "name" | "version">, path: string) {
    const provenance = this.get(name);
    if (!provenance) {
      const purl = typeof name === "string" ? name : toPurl(name);
      throw new Error(`Provenance not found for ${purl}`);
    }
    await writeFile(path, JSON.stringify(provenance));
  }
  static async fromFile(path: string) {
    const bundle = new ProvenanceBundle();
    const content = await readFile(path, "utf-8");
    try {
      const maybeSingle = JSON.parse(content);
      if (maybeSingle && typeof maybeSingle === "object") {
        bundle.add(maybeSingle);
      }
      return bundle;
    } catch {}
    const lines = content.split(/(\r\n|\n|\r)/g);
    for (const line of lines) {
      try {
        const provenance = JSON.parse(line);
        bundle.add(provenance);
      } catch {}
    }
    return bundle;
  }
}
