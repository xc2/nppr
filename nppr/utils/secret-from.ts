import { readFile, stat } from "node:fs/promises";

export class UnknownSecretType extends Error {
  constructor(type: string) {
    super(`Unknown secret type: ${type}`);
  }
}

export async function secretFrom(id: string) {
  const [type, ...rest] = id.split(":");
  if (type === "env") {
    return envSecret(rest.join(":"));
  } else if (type === "file") {
    return fileSecret(rest.join(":"));
  }
  throw new UnknownSecretType(type);
}

function envSecret(key: string) {
  return process.env[key];
}

async function fileSecret(path: string) {
  const { mode } = await stat(path);
  if ((mode & 0o0077) !== 0) {
    process.emitWarning(
      `Secret file \`${path}\` is too open, please consider changing the permissions`
    );
  }
  return (await readFile(path, "utf-8")).trim();
}
