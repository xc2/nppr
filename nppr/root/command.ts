import { writeFile } from "node:fs/promises";
import { glob } from "glob";
import { tryToNumber } from "nppr-core";
import { attest } from "nppr-core/provenance";
import { type PublishOptions, publish } from "nppr-core/publish";
import { repack } from "nppr-core/repack";
import { ArgumentsError, type CliCommand } from "../utils/cac";
import { Package } from "../utils/package";

export const rootCommand: CliCommand = (cmd) => {
  return cmd("[...inputs]")
    .option("--repack", "[Repack] Repack the tarball")
    .option(
      "--output [filepath]",
      "[Repack] Save the repacked tarball to a file, by default nothing will be written"
    )
    .option("--name <name>", "[Repack] Overrides `package.json/name` on repacking")
    .option("--version <version>", "[Repack] Overrides `package.json/version` on repacking")
    .option("--remap <remap>", "[Repack] Remap dependencies/optionalDependencies on repacking")
    .option(
      "--provenance [filepath]",
      "[Provenance] Generate and attest the provenance of the package"
    )
    .option("--registry <url>", "Package Registry URL", { default: "https://registry.npmjs.org" })
    .option("--publish", "[Publish] Publish the package to the registry")
    .option(
      "--keep-fields [fields]",
      "[Publish] Don't remove meaningless fields from manifest on publish"
    )
    .option("--add-fields [fields]", "[Publish] Additional fields to add to the manifest")
    .example(
      (bin) =>
        `  Rename a package

    $ ${bin} --repack bar-0.0.0.tgz --name foo --output
  
  Re-version a package

    $ ${bin} --repack bar-0.0.0.tgz --version 0.1.1 --output
`
    )
    .action(async (inputs: string[], options) => {
      const _inputs = inputs.map(tryToNumber);
      const _paths = _inputs.filter((v) => typeof v === "string");
      const _fds = _inputs.filter((v) => typeof v === "number");
      const pkgs = ([] as (string | number)[])
        .concat(await glob(_paths, {}))
        .concat(_fds)
        .map((input) => new Package(input));

      if (pkgs.length === 0) {
        throw new ArgumentsError("No packages given");
      }

      const writings: PromiseLike<any>[] = [];

      // #region Repack
      if (options.repack) {
        repack(
          pkgs.map((v) => ({ source: v.source })),
          {
            name: options.name,
            version: options.version,
            remapDeps: options.remap,
          }
        ).forEach((output, i) => {
          pkgs[i].source = output;
        });
      }
      // #endregion
      // #region Output
      if (options.output) {
        for (const pkg of pkgs) {
          writings.push(
            pkg.output(
              typeof options.output === "string"
                ? options.output
                : "[path]/[name]-[version]_repacked[extname]"
            )
          );
        }
      }
      // #endregion

      let provenance: any;

      // #region Provenance
      if (options.provenance) {
        provenance = await attest(
          pkgs.map((v) => ({ source: v.tee(), manifest: v.manifest })),
          {}
        );
        if (typeof options.provenance === "string") {
          writings.push(writeFile(options.provenance, JSON.stringify(provenance)));
        }
      }
      // #endregion

      // #region Publish
      if (options.publish) {
        const publishOptions: PublishOptions = {
          registry: options.registry,
          provenanceBundle: provenance,
          manifest: {
            keepFields: options.keepFields,
            additionalFields: options.addFields,
          },
        };
        for (const pkg of pkgs) {
          await publish(pkg.source, publishOptions);
        }
      }
      // #endregion

      await Promise.all(writings);
    });
};
