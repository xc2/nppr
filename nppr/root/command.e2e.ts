import { link, readFile } from "node:fs/promises";
import { mockFulcio, mockRekor, mockTSA } from "@sigstore/mock";
import { cac } from "cac";
import nock from "nock";
import { getManifest, inputSource } from "nppr-core";
import { generateSubject } from "nppr-core/provenance";
import { stubEnvs, test } from "tests/vitest";
import { afterAll, afterEach, beforeAll, beforeEach, describe, expect } from "vitest";
import { BasicTarballPath, ReadmeTemplate } from "../../tests/__fixtures__/tarball";
import { entryText } from "../../tests/helpers/iterator";
import { simpleJWT } from "../../tests/helpers/jwt";
import { registerCommand } from "../utils/cac";
import { rootCommand } from "./command";

const cwd = process.cwd();
beforeAll(() => {
  process.chdir(new URL("../testing", import.meta.url).pathname);
});
afterAll(() => {
  process.chdir(cwd);
});

beforeAll(() => {
  nock.disableNetConnect();
  // nock.enableNetConnect("tuf-repo-cdn.sigstore.dev");
});
afterAll(() => {
  nock.enableNetConnect();
});
beforeEach(() => {
  mockFulcio();
  mockRekor();
  mockTSA();
});
afterEach(() => {
  nock.cleanAll();
});

async function runCli(args: string[]) {
  const cli = cac("nppr");

  registerCommand(cli, "", rootCommand);

  cli.parse([process.argv[0], "nppr", ...args], { run: false });
  return cli.runMatchedCommand();
}

test("provenance", async ({ file, signal }) => {
  stubEnvs({ SIGSTORE_ID_TOKEN: simpleJWT({ sub: "foo", name: "bar" }) }, signal);
  const sigstore = file("sigstore.json");
  const subject = await generateSubject(BasicTarballPath);
  await runCli(["--provenance", sigstore, BasicTarballPath]);
  const bundle = JSON.parse(await readFile(sigstore, "utf8"));
  const payload = JSON.parse(Buffer.from(bundle.dsseEnvelope.payload, "base64").toString("utf-8"));

  expect(payload).toMatchObject({
    subject: expect.arrayContaining([subject]),
  });
});

describe("repack", () => {
  test("output to file", async ({ file }) => {
    const tarball = file("repacked.tgz");
    await runCli(["--repack", "--output", tarball, BasicTarballPath]);
    const manifest = await getManifest(inputSource(tarball));
    expect(manifest).toMatchObject({
      name: "barhop",
      version: "0.0.0-PLACEHOLDER",
    });
  });
  test("readme file", async ({ file }) => {
    const tarball = file("repacked.tgz");
    await runCli([
      "--repack",
      "--name",
      "foo",
      "--version",
      "1.0.0",
      "--readme",
      ReadmeTemplate,
      "--output",
      tarball,
      BasicTarballPath,
    ]);
    await expect(entryText(tarball, "README.md")).resolves.toMatchSnapshot();
  });
  test("rename, reversion, output to file with default name", async ({ file }) => {
    const source = file("source.tgz");
    await link(BasicTarballPath, source);
    const scope = "foo";
    const unscoped = Math.random().toString(36).slice(2);
    const name = `@${scope}/${unscoped}`;
    const version = [0, 0, 0].map((_) => Math.round(Math.random() * 10)).join(".");
    const tarball = file(`${scope}-${unscoped}-${version}_repacked.tgz`);
    await runCli(["--repack", "--name", name, "--version", version, source, "--output"]);
    const manifest = await getManifest(inputSource(tarball));
    expect(manifest).toMatchObject({
      name,
      version,
    });
  });
});

describe("publish", () => {
  const registry = "http://foo.internal:8080";
  const nockScope = nock(registry);
  const packageName = `pkg-not-exists-${Math.random().toString(36).slice(2, 6)}`;

  test("basic", async () => {
    nockScope.put(`/barhop`).reply(async (uri, body) => {
      expect(body).toMatchObject({
        name: "barhop",
        versions: {
          "0.0.0-PLACEHOLDER": expect.not.objectContaining({
            packageManager: expect.any(String),
          }),
        },
      });
      return [201];
    });
    expect(nockScope.isDone()).toBeFalsy();

    await runCli(["--publish", "--registry", registry, BasicTarballPath]);
    expect(nockScope.isDone()).toBeTruthy();
  });
  test("repack and publish", async () => {
    nockScope.put(`/${packageName}`).reply(async (uri, body) => {
      expect(body).toMatchObject({ name: packageName });
      return [201];
    });
    expect(nockScope.isDone()).toBeFalsy();

    await runCli([
      "--repack",
      "--name",
      packageName,
      "--publish",
      "--registry",
      registry,
      BasicTarballPath,
    ]);
    expect(nockScope.isDone()).toBeTruthy();
  });
  test("repack, provenance, and publish", async ({ signal }) => {
    stubEnvs({ SIGSTORE_ID_TOKEN: simpleJWT({ sub: "foo", name: "bar" }) }, signal);

    nockScope.put(`/${packageName}`).reply(async (uri, body) => {
      expect(body).toMatchObject({
        name: packageName,
        versions: {
          "0.0.0-PLACEHOLDER": {
            homepage: "https://github.com/xc2/barhop",
            gitHead: "3662536f93fb61b95e9a6c7106ba4bcfa2694115",
          },
        },
        _attachments: {
          [`${packageName}-0.0.0-PLACEHOLDER.sigstore`]: {
            data: expect.any(String),
          },
        },
      });
      return [201];
    });
    expect(nockScope.isDone()).toBeFalsy();

    await runCli([
      "--repack",
      "--name",
      packageName,
      "--provenance",
      "--publish",
      "--registry",
      registry,
      BasicTarballPath,
    ]);
    expect(nockScope.isDone()).toBeTruthy();
  });
  test("keep fields", async () => {
    nockScope.put(`/barhop`).reply(async (uri, body) => {
      expect(body).toMatchObject({
        name: "barhop",
        versions: {
          "0.0.0-PLACEHOLDER": {
            packageManager: expect.any(String),
          },
        },
      });
      return [201];
    });
    expect(nockScope.isDone()).toBeFalsy();

    await runCli([BasicTarballPath, "--publish", "--registry", registry, "--keep-fields"]);
    expect(nockScope.isDone()).toBeTruthy();
  });
  test("add fields", async () => {
    nockScope.put(`/barhop`).reply(async (uri, body) => {
      expect(body).toMatchObject({
        name: "barhop",
        versions: {
          "0.0.0-PLACEHOLDER": {
            readmeFilename: "PACKAGE.md",
          },
        },
      });
      return [201];
    });
    expect(nockScope.isDone()).toBeFalsy();

    await runCli([
      BasicTarballPath,
      "--publish",
      "--registry",
      registry,
      "--add-fields.readmeFilename",
      "PACKAGE.md",
    ]);
    expect(nockScope.isDone()).toBeTruthy();
  });
});
