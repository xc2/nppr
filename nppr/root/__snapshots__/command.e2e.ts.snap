// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`repack > readme file 1`] = `
"# foo

## Installation

\`\`\`bash
npm install foo@1.0.0
\`\`\`"
`;
